### ISSUE TITLE
TCP_BACKLOG runtime config option

### ISSUE THREAD

**Issue by SvenVD-be** [NLnetLabs/nsd#137]
Is there any way to increase the default TCP_BACKLOG option of nsd from 256 to higher values without recompiling and setting the value higher in AC_DEFINE_UNQUOTED([TCP_BACKLOG], [256]?

It would be nice if this was a configuration option like the tcp-listen-queue that bind has.

Thanks

**Comment by wtoorop** [comment]
Thanks @SvenVD-be . We'll consider (I'll bring it up internally).

**Comment by crrodriguez** [comment]
On Linux you should default to -1 (meaning OS default or sysadmin choosen value) not to an arbitrary number you picked.

**Comment by crrodriguez** [comment]
see also https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=19f92a030ca6d772ab44b22ee6a01378a8cb32d4

**Comment by SvenVD-be** [comment]
any update?

**Comment by k0ekk0ek** [comment]
Hi @SvenVD-be! Sorry for the overdue reply. Looking at the code, there are no changes yet. I don't know what was decided at the time. @wtoorop or @wcawijngaards any comments? If there's no objections, I guess it isn't too much work to add the option and the `-1` option indicated by @crrodriguez may actually be a good default, at least on Linux.

**Comment by wcawijngaards** [comment]
I do not recall, @k0ekk0ek . I also have no objections, for this option. Most DNS authority servers have very little TCP traffic, because they use UDP.

**Comment by SvenVD-be** [comment]
any update?

**Comment by wcawijngaards** [comment]
The commit adds the option `tcp-listen-queue: <number>`. It defaults to the same as before, with default 256. It is possible to set it to '-1', and then this is passed to the 'listen' call for backlog queue length. The metrics and TCP used by nsd-control are set at their own value, and the commit does not change those.

**Comment by crrodriguez** [comment]
> The commit adds the option `tcp-listen-queue: <number>`. It defaults to the same as before, with default 256. It is possible to set it to '-1', and then this is passed to the 'listen' call for backlog queue length. The metrics and TCP used by nsd-control are set at their own value, and the commit does not change those.

I strongly suggest you make it default to -1 in bsds and linux at least it is the system administrator or the kernel that must set a default and not the application..this as far as I know even has security implications. 

"Raising SOMAXCONN lowers chance of the port being unavailable under
even small SYNFLOOD attack, and reduces possibilities of side channel
vulnerabilities."

**Comment by wcawijngaards** [comment]
Certainly, the commit changes the default for `tcp-listen-queue` to `-1`, on `*bsd*` and `*linux*` systems. Hopefully that can provide better mitigation for those issues.
